nixos/grub: Name initrd-secrets by system, not by initrd

pwaller opened the pull request:
Previously, secrets were named according to the initrd they were
associated with. This created a problem: If secrets were changed whilst
the initrd remained the same, there were two versions of the secrets
with one initrd. The result was that only one version of the secrets would
by recorded into the /boot partition and get used. AFAICT this would
only be the oldest version of the secrets for the given initrd version.

This manifests as #114594, which I found frustrating while trying to use
initrd secrets for the first time. While developing the secrets I found
I could not get new versions of the secrets to take effect.
Additionally, it's a nasty issue to run into if you had cause to change
the initrd secrets for credential rotation, etc, if you change them and
discover you cannot, or alternatively that you can't roll back as you
would expect.

Additional changes in this patch.

* Add a regression test that switching to another grub configuration
  with the alternate secrets works. This test relies on the fact that it
  is not changing the initrd. I have checked that the test fails if I
  undo my change.

* Persist the useBootLoader boot disk state, similarly to other state.
  * I had to do this, otherwise I could not find a route to testing the
    alternate boot configuration. I did attempt a few different ways of
    testing this, including directly running install-grub.pl, but what
    I've settled on is most like what a user would do and avoids
    depending on lots of internal details.
  * Making tests that test the boot are a bit tricky (see hibernate.nix
    and installer.nix for inspiration), I found that in addition to
    having to copy quite a bit of code I still couldn't get things to
    work as desired since the bootloader state was being clobbered.

My change to persist the useBootLoader state could break things,
conceptually. I need some help here discovering if that is the case,
possibly by letting this run through a staging CI if there is one.

Fixes #114594.

cc potential reviewers:

@lopsided98 (original implementer) @joachifm (original reviewer),
@wkennington (numerous fixes to grub-install.pl), @lheckemann (wrote
original secrets test).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pwaller committed:
nixos/grub: Name initrd-secrets by system, not by initrd

Previously, secrets were named according to the initrd they were
associated with. This created a problem: If secrets were changed whilst
the initrd remained the same, there were two versions of the secrets
with one initrd. The result was that only one version of the secrets would
by recorded into the /boot partition and get used. AFAICT this would
only be the oldest version of the secrets for the given initrd version.

This manifests as #114594, which I found frustrating while trying to use
initrd secrets for the first time. While developing the secrets I found
I could not get new versions of the secrets to take effect.
Additionally, it's a nasty issue to run into if you had cause to change
the initrd secrets for credential rotation, etc, if you change them and
discover you cannot, or alternatively that you can't roll back as you
would expect.

Additional changes in this patch.

* Add a regression test that switching to another grub configuration
  with the alternate secrets works. This test relies on the fact that it
  is not changing the initrd. I have checked that the test fails if I
  undo my change.

* Persist the useBootLoader disk state, similarly to other boot state.
  * I had to do this, otherwise I could not find a route to testing the
    alternate boot configuration. I did attempt a few different ways of
    testing this, including directly running install-grub.pl, but what
    I've settled on is most like what a user would do and avoids
    depending on lots of internal details.
  * Making tests that test the boot are a bit tricky (see hibernate.nix
    and installer.nix for inspiration), I found that in addition to
    having to copy quite a bit of code I still couldn't get things to
    work as desired since the bootloader state was being clobbered.

My change to persist the useBootLoader state could break things,
conceptually. I need some help here discovering if that is the case,
possibly by letting this run through a staging CI if there is one.

Fix #114594.

cc potential reviewers:

@lopsided98 (original implementer) @joachifm (original reviewer),
@wkennington (numerous fixes to grub-install.pl), @lheckemann (wrote
original secrets test).

pwaller commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -152,9 +152,11 @@ let
+          # Create a writable copy/snapshot of the boot disk.
+          # A writable boot disk can be booted from automatically.
+          ${qemu}/bin/qemu-img create -f qcow2 -F qcow2 -b ${bootDisk}/disk.img "$TMPDIR/disk.img"
+        fi
```
I could use a hand determining if this breaks anything. It's necessary in order for `grub-reboot` to take effect. I could see it breaking things if there were tests existing that expected this to be fresh state every time. My view is, if there are not such tests, this makes it more consistent with the rest of the script (which uses `! test -e` to choose whether to create the other files).

I would like to know if we can throw this at a CI somewhere and see if it breaks any tests, alternatively, if there is a set of tests I can reasonably run locally which would be considered good enough.

pwaller replied:
In case anyone is wondering how it works for the installer and hibernate tests and why I shouldn't do the same here: those tests do a complete install of nixos, and involves a fair amount of extra logic that I would prefer to avoid.

rnhmjoj replied:
I really don't know if this is something we want to do, by default or not at all.
To play safe I would add a `persistBootLoader` option to toggle the behavior and use only where it's needed.

pwaller replied:
Done.

JohannesEbke commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -450,7 +450,8 @@ sub addEntry {
     if (-e -x "$path/append-initrd-secrets") {
-        my $initrdName = basename($initrd);
+        # Name the initrd secrets after the system from which they're derived.
+        my $initrdName = basename(Cwd::abs_path("$path"));
```
Maybe this variable should now get its name changed, as it is no longer the initrd name but the system name?

pwaller replied:
Good point, I'm updating this.

JohannesEbke commented on nixos/tests/initrd-secrets-changing.nix:
```diff
@@ -0,0 +1,55 @@
+        "cmp ${secret1InStore} /secret-from-initramfs",
+        "cmp ${secret1InStore} /run/keys/test",
+    )
+    machine.succeed("grub-reboot 1")
```
I had to read the test twice to spot that this reboots into the specialisation 2 (second entry in Grub, numbered from zero). 

According to the `grub-reboot` man page this could also be a menu item title - more readable, but probably unpredictable?

I think at least a comment would be nice, though!

pwaller replied:
For this I copied the pattern from the nixos-installer tests:

https://github.com/NixOS/nixpkgs/blob/af0b453724a59bc8fa90fa1dfad30896cc3df529/nixos/tests/installer.nix#L243-L244

You're right that I predicting the correct boot string is harder. Doable, but I figured from looking at the `grub.cfg` and knowing how the `grub.cfg` is constructed that taking the second entry would be the more straightforward and reliable approach.

JohannesEbke reviewed (approved):
Looks good for me from the code side now.

rnhmjoj commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -450,8 +450,9 @@ sub addEntry {
-        my $initrdSecretsPath = "$bootPath/kernels/$initrdName-secrets";
+        # Name the initrd secrets after the system from which they're derived.
+        my $systemName = basename(Cwd::abs_path("$path"));
+        my $initrdSecretsPath = "$bootPath/kernels/$systemName-secrets";
```
This makes sense a lot of sense: it should have always been like this. 
I don't know why there's a `$initrdName` if the name is going to be `initrd` anyway... maybe it's not intentional.

pwaller commented:
* Rebased and per @rnhmjoj's comment, introduced `virtualisation.persistBootDevice` used to control the disk persistence, leaving the default as false.

rnhmjoj commented:
We should do the same change in systemd-boot, which is the only other bootloader that sets `supportsInitrdSecrets = true`. The script is in
```
nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py
```
